crates/tarball/src/lib.rs

```diff
@@ -128,24 +128,23 @@ pub async fn download_tarball_to_store(
             }
             CacheValue::InProgress(notify) => Arc::clone(notify),
         };
+
+        tracing::info!(target: "pacquet::download", ?package_url, "Wait for cache");
         notify.notified().await;
-        if let Some(cached) = cache.get(package_url) {
-            if let CacheValue::Available(cas_paths) = &*cached.read().await {
-                return Ok(Arc::clone(cas_paths));
-            }
+        if let CacheValue::Available(cas_paths) = &*cache_lock.read().await {
+            return Ok(Arc::clone(cas_paths));
         }
-        Err(TarballError::Io(std::io::Error::new(
-            std::io::ErrorKind::Other,
-            "Failed to get or compute tarball data",
-        )))
+        unreachable!("Failed to get or compute tarball data for {package_url:?}");
     } else {
         let notify = Arc::new(Notify::new());
         let cache_lock = notify
             .pipe_ref(Arc::clone)
             .pipe(CacheValue::InProgress)
             .pipe(RwLock::new)
             .pipe(Arc::new);
-        cache.insert(package_url.to_string(), Arc::clone(&cache_lock));
+        if cache.insert(package_url.to_string(), Arc::clone(&cache_lock)).is_some() {
+            tracing::warn!(target: "pacquet::download", ?package_url, "Race condition detected when writing to cache");
```

Comment:
Is there no way to exclude the possibility of a race condition?

Reply:
No, I don't think so. What is the worst that could happen? We would download the same package twice? That doesn't sound like a big issue.

Reply:
OK. So do you want to leave it as a warning or remove the warning?

Reply:
It is fine with the warning. I was just wondering if Rust had a way to avoid this issue altogether. I guess you could use a semaphore with 1 permit between getting and setting? But that might be overengineering.

Reply:
> I was just wondering if Rust had a way to avoid this issue altogether.

The Rust type system is strong but not that strong.

> I guess you could use a semaphore with 1 permit between getting and setting?

Also known as a `notify`, which is already in used.

This warning is but a substitute for the inability of the Rust type system to prove that the key is not already set. Similar to how `unreachable!` is used to substitute Rust's inability to prove unreachable code paths.

Reply:
anyway, it can be left as it is now.